express-notes-api/server.js

```diff
@@ -0,0 +1,117 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+
+const app = express();
+
+app.use(express.json());
+
+// gets a list of notes in an array
+app.get('/api/notes', async (req, res) => {
+  const data = await fetchFile();
+  const array = [];
+
+  for (const note in data.notes) {
+    array.push(data.notes[note]);
+  }
+  res.status(200).send(array);
+});
+
+// gets a single note by id
+app.get('/api/notes/:id', async (req, res) => {
+  const data = await fetchFile();
+  const id = req.params.id;
+
+  idErrorHandler(id, data, res);
+  res.status(200).send(data.notes[id]);
+});
+
+// post a new note
+app.post('/api/notes', async (req, res) => {
+  const data = await fetchFile();
+  const body = req.body;
+  const id = data.nextId;
+
+  if (Object.keys(body).length === 0) {
+    res.status(400).send({ error: 'content is a required field' });
+    return;
+  }
+
+  try {
+    body.id = data.nextId;
+    data.notes[id] = body;
+    data.nextId++;
+    await writeFileDataJson(data);
+    res.status(201).send(body);
+  } catch (error) {
+    console.error(error);
+    res.status(500).send({ error: 'An unexpected error occurred.' });
+  }
+});
+
+// delete note
+app.delete('/api/notes/:id', async (req, res) => {
+  const data = await fetchFile();
+  const id = req.params.id;
+
+  idErrorHandler(id, data, res);
+  try {
+    delete data.notes[id];
+    await writeFileDataJson(data);
+    res.sendStatus(204);
+  } catch (error) {
+    console.error(error);
+    res.status(500).send({ error: 'An unexpected error occurred.' });
+  }
+});
+
+// replace a note
+app.put('/api/notes/:id', async (req, res) => {
+  const data = await fetchFile();
+  const id = req.params.id;
+  const body = req.body;
+
+  idErrorHandler(id, data, res);
+  try {
+    body.id = id;
+    data.notes[id] = body;
+    await writeFileDataJson(data);
+    res.status(200).send(body);
```

Comment:
Express sees that `body` is an object, so it sends JSON. However, it's more clear to explicitly use `.json` to make sure it always sends JSON.